---
title: "urti2021"
output: html_document

---

          
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```


```{r cars}
library(vegan)
library(picante)
library(phyloseq)
library(plyr)
library(ggplot2)
library(plyr)
library(reshape2)
library(patchwork) #拼图
library(cowplot)
library(ggpubr) #统计
library(GUniFrac)
library(microbiome)
suppressWarnings(suppressMessages(library(amplicon)))
```

## make phyloseq of 16S
## 16s分析

```{r}
setwd("C:/Users/DELL/Desktop/URTI.16SV4.202110/qiime2out")

##构建对象

otu <- read.delim('feature-table.txt', row.names = 1, sep = '\t', stringsAsFactors = FALSE, check.names = FALSE)
metadata <- read.table('metadata.txt', header = T, sep = '\t', row.names = 1)
metadata <- metadata[colnames(otu),]
tax<- read.table('tax.txt', header = T, sep = '\t', row.names = 1)
tree<- read.tree('tree.nwk')
repseqFile = "rep.fasta"
rep.seq = Biostrings::readDNAStringSet(repseqFile)

otumat<- otu

OTU = otu_table(otumat, taxa_are_rows = TRUE)
taxmat<- as.matrix(tax)
TAX = tax_table(taxmat)
sampledata<- sample_data(metadata)


physeq0 = merge_phyloseq(OTU, TAX)
physeq = merge_phyloseq(physeq0, sampledata, tree, rep.seq)
physeq2 = merge_phyloseq(OTU, TAX, sampledata, tree, rep.seq)
identical(physeq, physeq2)

#过滤样本
sub_physeq <- subset_taxa(physeq, kindom == "k__Bacteria")
prevdf = apply(X = otu_table(sub_physeq),
               MARGIN = ifelse(taxa_are_rows(sub_physeq), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

physeq.great<- prune_taxa(names(which(prevdf >= 2)), sub_physeq)

set.seed(003)
physeq.r = rarefy_even_depth(physeq.great, sample.size =20000, replace = TRUE)

physeq.rp <- prune_taxa(taxa_sums(physeq.r) > 0, physeq.r)
```


# 物种组成

```{r}
# phylum水平
cols15 = c("#660066", "#56B4E9", "#009E73", "#FF0000","#333333", "#F0E442",  "#D55E00","#CC79A7","#990000","#9900cc","#66FF66","#663300","#0000FF","#CC0033","#999999")

phylum.p = tax_glom(physeq.r2, "phylum")
otus <- as.data.frame(phylum.p@otu_table@.Data)
tax <- as.data.frame(phylum.p@tax_table@.Data)
tax1 <- tax[!duplicated(tax$phylum),]
otus1 <- otus[rownames(tax1),]
rownames(otus1) <- tax1$phylum
Unknown <- colSums(physeq.r2@otu_table@.Data) - colSums(otus1)
otus2 <- rbind(otus1, Unknown)
tax_stackplot(otus2, meta, groupID="groups2", topN=15, style="sample")+ scale_fill_manual(values=cols15)

cols20 = c("#FF0099", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2","#D55E00","#CC79A7","#990000","#9900cc","#66FF66","#663300","#0000FF","#CC0033","#FF0000","#000099","#660066","#333333","#9966CC","#999999")
#genus水平
genus.p = tax_glom(physeq.r2, "genus")
otus <- as.data.frame(genus.p@otu_table@.Data)
tax <- as.data.frame(genus.p@tax_table@.Data)
tax1 <- tax[!duplicated(tax$genus),]
otus1 <- otus[rownames(tax1),]
rownames(otus1) <- tax1$genus
Unknown <- colSums(physeq.r2@otu_table@.Data) - colSums(otus1)
otus2 <- rbind(otus1, Unknown)
tax_stackplot(otus2, meta, groupID="groups2", topN=20, style="sample")+ scale_fill_manual(values=cols20)

physeq.r3<- prune_taxa(names(which(prevdf >= 12)), physeq.r2)
genus.p = tax_glom(physeq.r3, "genus")
otus <- as.data.frame(genus.p@otu_table@.Data)
tax <- as.data.frame(genus.p@tax_table@.Data)
tax1 <- tax[!duplicated(tax$genus),]
otus1 <- otus[rownames(tax1),]
rownames(otus1) <- tax1$genus

otus2<- log10((otus1+0.1)/5000)
annotation_col = data.frame(
  group1 = meta$groups1, 
  group2 = meta$groups2
)
rownames(annotation_col) = colnames(otus2)
head(annotation_col)
annotation_row = data.frame(
  phylum = tax1$phylum
)
rownames(annotation_row) = rownames(otus2)
head(annotation_col)
pheatmap(otus2, annotation_col = annotation_col, annotation_row = annotation_row)

```

## alpha多样性

```{r}
#load('physeq.rp.Rdata')
col2 <- c("#ad1500", "#2B6B8E")
col3 <- c("#EC3D33", "#e3766b","#2B6B8E")
col5<- c("#824533", "#B79570","#CAD09E", "#FFCCB1", "#2B6B8E")

s.sh <- plot_richness(physeq.rp, "group",  measures=NULL)
s.data <- s.sh$data

s.data1 <- subset(s.data, group != "NA" & variable %in% c("Observed"))
s.data1$group1 <- factor(s.data1$group1,levels=c("infection", "control"))

my_comparisons2 <- list(c("infection", "control"))
pa1 <- ggplot(data=s.data1, aes(x = group1, y = value, group=group1)) +
    geom_boxplot(aes(x=group1, fill=group1), notch=FALSE, outlier.colour = "grey") +
    stat_summary(aes(x=group1), fun.y=mean, geom="point", color="grey") +
    geom_jitter(width=0.2, size=1.5, alpha = 0.75) +
    scale_fill_manual(values=col2) +
    theme(title=element_text(size=14,color="#4F4F4F"))+
    #facet_wrap(.~variable, scales = "free", nrow=1)+
    stat_compare_means(comparisons = my_comparisons2, paired = FALSE, color="black")+
    theme_classic()+ 
    theme(axis.text.x = element_text(size=14),axis.text.y = element_text(size=14)) + 
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
    theme(legend.title = element_blank())+ 
    theme(legend.position='none')+
    theme(plot.title = element_text(hjust = 0.5))
# +theme(strip.text = element_text(colour = 'black', face = 'bold', size = rel(1.2)), strip.background = element_rect(fill = 'white', colour = 'black', size = rel(2), linetype = 1))

s.data1$group2 <- factor(s.data1$group2,levels=c("infection", "co-infection", "control"))
my_comparisons3 <- list(c("co-infection", "control"), c("infection", "control"))
pa2 <- ggplot(data=s.data1, aes(x = group2, y = value, group=group2)) +
    geom_boxplot(aes(x=group2, fill=group2), notch=FALSE, outlier.colour = "grey") +
    stat_summary(aes(x=group2), fun.y=mean, geom="point", color="grey") +
    geom_jitter(width=0.2, size=1.5, alpha = 0.75) +
    scale_fill_manual(values=col3) +
    theme(title=element_text(size=14,color="#4F4F4F"))+
    #facet_wrap(.~variable, scales = "free", nrow=1)+
    stat_compare_means(comparisons = my_comparisons3, paired = FALSE, color="black")+
    theme_classic()+ 
    theme(axis.text.x = element_text(size=14),axis.text.y = element_text(size=14)) + 
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
    theme(legend.title = element_blank())+ 
    theme(legend.position='none')+
    theme(plot.title = element_text(hjust = 0.5))

sub_s.data <- subset(s.data, group %in% c("CTL", "A", "B", "S", "H") & variable %in% c("Observed"))
sub_s.data$group <- factor(sub_s.data$group,levels=c("A", "B", "S", "H", "CTL"))
my_comparisons5 <- list(c("CTL", "H"), c("CTL", "S"), c("CTL", "B"), c("A", "CTL"))
pa3 <- ggplot(data=sub_s.data, aes(x = group, y = value, group=group)) +
    geom_boxplot(aes(x=group, fill=group), notch=FALSE, outlier.colour = "grey") +
    stat_summary(aes(x=group), fun.y=mean, geom="point", color="grey") +
    geom_jitter(width=0.2, size=1.5, alpha = 0.75) +
    scale_fill_manual(values=col5) +
    theme(title=element_text(size=14,color="#4F4F4F"))+
    #facet_wrap(.~variable, scales = "free", nrow=1)+
    stat_compare_means(comparisons = my_comparisons5, paired = FALSE, color="black")+
    theme_classic()+ 
    theme(axis.text.x = element_text(size=14),axis.text.y = element_text(size=14)) + 
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
    theme(legend.title = element_blank())+ 
    theme(legend.position='none')+
    theme(plot.title = element_text(hjust = 0.5))

(pa1 + pa2 + pa3 + plot_layout(nrow = 1, widths = c(2.3, 3.2, 5)))


s.data1 <- subset(s.data, group == "A" & variable %in% c("Observed"))

p1<- ggplot(data=s.data1 , aes( x = value, y = log(FluA))) +
    geom_point(alpha=0.6, color="black",size=2)+ 
    geom_smooth(method=lm, size=2, color="#824533")+
    theme_bw()+
    theme(axis.text.x = element_text(size=14), axis.text.y = element_text(size=14)) +
    theme(legend.title = element_blank())+ 
    theme(title=element_text(size=11))+
    theme(legend.position="none")+
    stat_cor(method = "spearman", color="black", size=6)

s.data1 <- subset(s.data, group == "B" & variable %in% c("Observed"))

p2<- ggplot(data=s.data1 , aes( x = value, y = log(FluB))) +
    geom_point(alpha=0.6, color="black",size=2)+ 
    geom_smooth(method=lm, size=2, color="#B79570")+
    theme_bw()+
    theme(axis.text.x = element_text(size=14), axis.text.y = element_text(size=14)) +
    theme(legend.title = element_blank())+ 
    theme(title=element_text(size=11))+
    theme(legend.position="none")+
    stat_cor(method = "spearman", color= "black", size=6)

s.data1 <- subset(s.data, group == "S" & variable %in% c("Observed"))

p3<- ggplot(data=s.data1 , aes( x = value, y = log(RSV))) +
    geom_point(alpha=0.6, color="black",size=2)+ 
    geom_smooth(method=lm, size=2, color="#CAD09E")+
    theme_bw()+
    theme(axis.text.x = element_text(size=14), axis.text.y = element_text(size=14)) +
    theme(legend.title = element_blank())+ 
    theme(title=element_text(size=11))+
    theme(legend.position="none")+
    stat_cor(method = "spearman", color= "black", size=6)

s.data1 <- subset(s.data, group == "H" & variable %in% c("Observed"))

p4<- ggplot(data=s.data1 , aes( x = value, y = log(RHV))) +
    geom_point(alpha=0.6, color="black",size=2)+ 
    geom_smooth(method=lm, size=2, color="#FFCCB1")+
    theme_bw()+
    theme(axis.text.x = element_text(size=14), axis.text.y = element_text(size=14)) +
    theme(legend.title = element_blank())+ 
    theme(title=element_text(size=11))+
    theme(legend.position="none")+
    stat_cor(method = "spearman", color= "black", size=6)

(p1 | p2) / (p3 | p4)

library(tidyverse)
library("randomForest")
library("rfUtilities")
library("rfPermute")

physeq.a<- subset_samples(physeq.rp, group == "A")
physeq.ap<-  prune_taxa(taxa_sums(physeq.a) > 0, physeq.a)
physeq.ap.genus = tax_glom(physeq.ap, "genus")
otus <- as.data.frame(t(physeq.ap.genus@otu_table@.Data))
colnames(otus) <- as.data.frame(physeq.ap.genus@tax_table@.Data)$genus
otus$Observed <- subset(s.data, group == "A" & variable %in% c("Observed"))$FluA
otus1 <- otus[,!colnames(otus) %in% c("g__uncultured", "g__uncultured bacterium", "g__unidentified")]
colnames(otus1) <- make.names(colnames(otus1))

set.seed(123)
richness_rf <- randomForest(Observed ~ ., data= otus1, importance=TRUE,proximity=TRUE)
set.seed(123)
richness_perm <- rf.significance(richness_rf, otus1[,-136], nperm=99, ntree=501)
set.seed(123)
richness_rfP<- rfPermute(Observed ~ ., data = otus1, ntree = 500,
                         na.action = na.omit, nrep = 500,num.cores = 6
                        )
plotImportance(richness_rfP)
confusionMatrix(rp)

names <- c("g__Corynebacterium", "g__Ruminococcaceae.UCG.014", "g__Cardiobacterium", "g__uncultured.Candidatus.Saccharibacteria.bacterium", "g__Tannerella", "g__Porphyromonas", "g__Prevotella.1", "g__Olsenella", "g__Acholeplasma")
otus2 <- otus1[,names]
otus2$FluA <- subset(s.data, group == "A" & variable %in% c("Observed"))$FluA
otus3 <- melt(data.frame(otus2), id="FluA")
ggplot(data=otus3 , aes( y = log10(FluA), x = value/200)) +
    geom_point(alpha=0.6, color="black",size=2)+ 
    geom_smooth(method=lm, size=2, color="black")+
    facet_wrap(.~variable, scales = "free", nrow=3)+
    theme_classic()+
    theme(axis.text.x = element_text(size=14), axis.text.y = element_text(size=14)) +
    theme(legend.title = element_blank())+ 
    theme(title=element_text(size=11))+
    theme(legend.position="none")+
    stat_cor(method = "spearman", color= "black", size=6)

```

#beta多样性

```{r}
col5<- c("#824533", "#B79570","#CAD09E", "#FFCCB1", "#2B6B8E")

#FluA
physeq.rp5<- subset_samples(physeq.rp, group %in% c("CTL", "A"))
physeq5<-  prune_taxa(taxa_sums(physeq.rp5) > 0, physeq.rp5)

GP.ord <- ordinate(physeq5, "NMDS", "bray")
data1<- plot_ordination(physeq5, GP.ord, color="group")
data2 <- data1$data
ggscatter(data2, x= "NMDS1", y = "NMDS2", 
          color = "group", palette = c("#824533", "#2B6B8E"),
          ellipse = TRUE,  
          mean.point = FALSE, star.plot = FALSE,  
          ellipse.level = 0.3,  
          ggtheme = theme_minimal(),
          rug = FALSE)+
    theme(axis.title.x = element_text(size = 16,
                                      face = "bold", 
                                      vjust = 0.5, 
                                      hjust = 0.5))+
    theme(axis.title.y = element_text(size = 16,
                                      face = "bold", 
                                      vjust = 0.5, 
                                      hjust = 0.5))+
    theme_classic()

meta <- meta(physeq5)
dis_bray.f<- phyloseq::distance(physeq5, "bray")
set.seed(002)
adon.results<-adonis(dis_bray.f~ meta$group, perm=999)
knitr::kable(adon.results[["aov.tab"]])

#FluB
physeq.rp5<- subset_samples(physeq.rp, group %in% c("CTL", "B"))
physeq5<-  prune_taxa(taxa_sums(physeq.rp5) > 0, physeq.rp5)

GP.ord <- ordinate(physeq5, "NMDS", "bray")
data1<- plot_ordination(physeq5, GP.ord, color="group")
data2 <- data1$data
ggscatter(data2, x= "NMDS1", y = "NMDS2", 
          color = "group", palette = c("#B79570", "#2B6B8E"),
          ellipse = TRUE,  
          mean.point = FALSE, star.plot = FALSE,  
          ellipse.level = 0.3,  
          ggtheme = theme_minimal(),
          rug = FALSE)+
    theme(axis.title.x = element_text(size = 16,
                                      face = "bold", 
                                      vjust = 0.5, 
                                      hjust = 0.5))+
    theme(axis.title.y = element_text(size = 16,
                                      face = "bold", 
                                      vjust = 0.5, 
                                      hjust = 0.5))+
    theme_classic()

meta <- meta(physeq5)
dis_bray.f<- phyloseq::distance(physeq5, "bray")
set.seed(002)
adon.results<-adonis(dis_bray.f~ meta$group, perm=999)
knitr::kable(adon.results[["aov.tab"]])

#RSV
physeq.rp5<- subset_samples(physeq.rp, group %in% c("CTL", "S"))
physeq5<-  prune_taxa(taxa_sums(physeq.rp5) > 0, physeq.rp5)

GP.ord <- ordinate(physeq5, "NMDS", "bray")
data1<- plot_ordination(physeq5, GP.ord, color="group")
data2 <- data1$data
ggscatter(data2, x= "NMDS1", y = "NMDS2", 
          color = "group", palette = c("#2B6B8E", "#CAD09E"),
          ellipse = TRUE,  
          mean.point = FALSE, star.plot = FALSE,  
          ellipse.level = 0.3,  
          ggtheme = theme_minimal(),
          rug = FALSE)+
    theme(axis.title.x = element_text(size = 16,
                                      face = "bold", 
                                      vjust = 0.5, 
                                      hjust = 0.5))+
    theme(axis.title.y = element_text(size = 16,
                                      face = "bold", 
                                      vjust = 0.5, 
                                      hjust = 0.5))+
    theme_classic()

meta <- meta(physeq5)
dis_bray.f<- phyloseq::distance(physeq5, "bray")
set.seed(002)
adon.results<-adonis(dis_bray.f~ meta$group, perm=999)
knitr::kable(adon.results[["aov.tab"]])

# RHV
physeq.rp5<- subset_samples(physeq.rp, group %in% c("CTL", "H"))
physeq5<-  prune_taxa(taxa_sums(physeq.rp5) > 0, physeq.rp5)

GP.ord <- ordinate(physeq5, "NMDS", "bray")
data1<- plot_ordination(physeq5, GP.ord, color="group")
data2 <- data1$data
ggscatter(data2, x= "NMDS1", y = "NMDS2", 
          color = "group", palette = c( "#2B6B8E", "#FFCCB1"),
          ellipse = TRUE,  
          mean.point = FALSE, star.plot = FALSE,  
          ellipse.level = 0.3,  
          ggtheme = theme_minimal(),
          rug = FALSE)+
    theme(axis.title.x = element_text(size = 16,
                                      face = "bold", 
                                      vjust = 0.5, 
                                      hjust = 0.5))+
    theme(axis.title.y = element_text(size = 16,
                                      face = "bold", 
                                      vjust = 0.5, 
                                      hjust = 0.5))+
    theme_classic()

meta <- meta(physeq5)
dis_bray.f<- phyloseq::distance(physeq5, "bray")
set.seed(002)
adon.results<-adonis(dis_bray.f~ meta$group, perm=999)
knitr::kable(adon.results[["aov.tab"]])

```

# lefse + 物种箱线图
```{r}
physeq.rp5<- subset_samples(physeq.rp, group %in% c("CTL", "A", "B", "S", "H"))
physeq5<-  prune_taxa(taxa_sums(physeq.rp5) > 0, physeq.rp5)
physeq.5rel <- transform_sample_counts(physeq5, function(x)x / sum(x))
physeq5.genus = tax_glom(physeq.5rel, "genus")
otus <- as.data.frame(t(physeq5.genus@otu_table@.Data))
colnames(otus) <- as.data.frame(physeq5.genus@tax_table@.Data)$genus
otus$group <- as.data.frame(meta(physeq5.genus))$group
otus1 <- otus[,!colnames(otus) %in% c("g__uncultured", "g__uncultured bacterium", "g__uncultured rumen bacterium","g__uncultured forest soil bacterium" ,"g__hydrothermal vent metagenome" ,"g__uncultured soil bacterium" ,"g__uncultured compost bacterium" ,"g__unidentified")]

# write.csv(otus1, file="genus.lefse.csv")

library(ggthemes)
ggplot(lefse,aes(x = reorder(genus,flua.lda),y=rhv.lda,
              fill = rhv.group,
              label = rhv.lda))+
    geom_col(show.legend = FALSE) +
    coord_flip() +
    xlab('Gene Name') +
    ggtitle('Expression level of significant changed genes')+
    theme_clean()

data <- -lefse[,c(2,6,10,14)]
data1 <- (data-min(data))/(max(data)-min(data))
library(ggbiplot)
wine.pca <- prcomp(data1, scale. = TRUE)
ggbiplot(wine.pca, obs.scale = 1, var.scale = 1, ellipse = FALSE, circle = FALSE) +
    scale_color_discrete(name = '') +
    theme(legend.direction = 'horizontal', legend.position = 'top')+
    theme_classic()+
    geom_text_repel(aes(label=rownames(data1)))


otus1$group <- factor(otus1$group,levels=c("A", "B", "S", "H", "CTL"))
my_comparisons5 <- list(c("CTL", "H"), c("CTL", "S"), c("CTL", "B"), c("A", "CTL"))

ggplot(data=otus1, aes(x = group, y = 100*g__Granulicatella, group=group)) +
    geom_boxplot(aes(x=group, fill=group), notch=FALSE, outlier.colour = "grey") +
    stat_summary(aes(x=group), fun.y=mean, geom="point", color="grey") +
    geom_jitter(width=0.2, size=1.5, alpha = 0.75) +
    scale_fill_manual(values=col5) +
    theme(title=element_text(size=12))+
    #facet_wrap(.~variable, scales = "free", nrow=1)+
    stat_compare_means(comparisons = my_comparisons5, paired = FALSE, color="black")+
    theme_classic()+ 
    theme(axis.text.x = element_text(size=14),axis.text.y = element_text(size=14)) + 
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
    theme(legend.title = element_blank())+ 
    theme(legend.position='none')+
    theme(plot.title = element_text(hjust = 0.5))+
    labs(title= "Granulicatella", y="Realative Abundance", x="")+
	    scale_y_continuous(limits=c(0,25), breaks=seq(0,25,5))

ggplot(data=otus1, aes(x = group, y = 100*g__Veillonella, group=group)) +
    geom_boxplot(aes(x=group, fill=group), notch=FALSE, outlier.colour = "grey") +
    stat_summary(aes(x=group), fun.y=mean, geom="point", color="grey") +
    geom_jitter(width=0.2, size=1.5, alpha = 0.75) +
    scale_fill_manual(values=col5) +
    theme(title=element_text(size=12))+
    #facet_wrap(.~variable, scales = "free", nrow=1)+
    stat_compare_means(comparisons = my_comparisons5, paired = FALSE, color="black")+
    theme_classic()+ 
    theme(axis.text.x = element_text(size=14),axis.text.y = element_text(size=14)) + 
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
    theme(legend.title = element_blank())+ 
    theme(legend.position='none')+
    theme(plot.title = element_text(hjust = 0.5))+
    labs(title= "Veillonella", y="Realative Abundance", x="")


```

#预测模型

```{r}
library(pROC) 
library(randomForest) 
library(e1071)
library(glmnet)

physeq.rpa<- subset_samples(physeq.rp, group %in% c("A", "B", "S", "H", "AB", "ABS", "AS", "BH", "BS", "CTL"))

physeq.rpa<-  prune_taxa(taxa_sums(physeq.rpa) > 0, physeq.rpa)
physeq10 = tax_glom(physeq.rpa, "genus")
data <- as.data.frame(t(physeq10@otu_table@.Data))
colnames(data) <- as.data.frame(physeq10@tax_table@.Data)$genus
colnames(data) <- make.names(colnames(data))
data1 <- data[,!colnames(data) %in% c("g__uncultured", "g__uncultured bacterium", "g__uncultured rumen bacterium","g__uncultured forest soil bacterium" ,"g__hydrothermal vent metagenome" ,"g__uncultured soil bacterium" ,"g__uncultured compost bacterium" ,"g__unidentified")]

data1$group <- as.factor(as.data.frame(meta(physeq.rpa))$group1)
set.seed(123)
rf.fit<- randomForest(group ~ ., data = data1, importance = TRUE, )
plot.roc(data1$group, 
    rf.fit$votes[,1], 
    percent=TRUE,
    xlab="False Positive Percentage", 
    ylab="True Postive Percentage", col="#377eb8", lwd=4, print.auc=TRUE)
varImpPlot(rf.fit, main = "variable importance")

votes <- as.data.frame(rf2$votes)
votes$group <- data$group
longtest <- melt_roc(votes, "group", c("control", "infection", "co-infection"))

ggplot(longtest, aes(d = D, m = M, color = name)) + geom_roc() + style_roc()


data$group <- as.factor(as.data.frame(meta(physeq.rpa))$group2)
set.seed(123)
rf2<- randomForest(group ~ ., data = data, importance = TRUE, )



roc(group, 
    glm.fit$fitted.values, 
    plot=TRUE, 
    legacy.axes=TRUE, 
    percent=TRUE, 
    xlab="False Positive Percentage", 
    ylab="True Postive Percentage", col="#377eb8", lwd=4, print.auc=TRUE)

# 在以上基础上，添加svm模型的ROC
plot.roc(group,
         pred_2,
         percent=TRUE, 
         col="red", 
         lwd=4, 
         print.auc=TRUE, 
         add=TRUE, 
         print.auc.y=40)

# 在logistic回归模型的基础上，添加随机森林模型的ROC
roc(group, 
         rf.fit$votes[,1], 
         percent=TRUE, 
         col="#4daf4a", 
         lwd=4, 
         print.auc=TRUE, 
         add=TRUE, 
         print.auc.y=30, 
    xlab="False Positive Percentage", 
    ylab="True Postive Percentage", col="#377eb8", lwd=4, print.auc=TRUE)

# 设置图例标签
legend("bottomright", 
       legend=c("Logisitic Regression", "SVM" , "Random Forest"), 
       col=c("#377eb8", "red", "#4daf4a"), 
       lwd=4)


data <- otu.train[,c(1,389:529)]
smp.size = floor(0.7*nrow(data)) 
set.seed(123456789)                     
train.ind = sample(seq_len(nrow(data)), smp.size)
train = data[train.ind, ] 
test = data[-train.ind, ] 

model = svm(formula = groups2 ~ ., data = train,type = 'C',kernel = "linear")
train.pred = predict(model, train)
test.pred = predict(model, test)

table(real=train$groups2, predict=train.pred) 

confus.matrix = table(real=train$groups2, predict=train.pred)
sum(diag(confus.matrix))/sum(confus.matrix)

## 测试集的混淆矩阵
table(real=test$groups2, predict=test.pred)

confus.matrix = table(real=test$groups2, predict=test.pred)
sum(diag(confus.matrix))/sum(confus.matrix)



# Health vs Infection
physeq.rpa<- subset_samples(physeq.rp, group != "NA")
physeq.rpa<-  prune_taxa(taxa_sums(physeq.rpa) > 0, physeq.rpa)
physeq10 = tax_glom(physeq.rpa, "genus")
data <- as.data.frame(t(physeq10@otu_table@.Data))
colnames(data) <- as.data.frame(physeq10@tax_table@.Data)$genus
colnames(data) <- make.names(colnames(data))
data.g <- data[,!colnames(data) %in% c("g__uncultured", "g__uncultured bacterium", "g__uncultured rumen bacterium","g__uncultured forest soil bacterium" ,"g__hydrothermal vent metagenome" ,"g__uncultured soil bacterium" ,"g__uncultured compost bacterium" ,"g__unidentified")]

colnames(data) <- make.names(colnames(data))

meta <- as.data.frame(meta(physeq.rpa))
group1 <- subset(meta, group1 == "control")
group2 <- subset(meta, group1 != "control")
data1 <- data.g[rownames(group1),]
data2 <- data.g[sample(rownames(group2), nrow(group1)),]
data1$group <- "Health"
data2$group <- "Infection"
data.all <- rbind(data1, data2)
data.all$group <- as.factor(data.all$group)

set.seed(123)
richness_rfP<- rfPermute(group ~ ., data = data.all, ntree = 500,
                         na.action = na.omit, nrep = 100)
roc <- roc(data.all$group, richness_rfP$rf$votes[,1], percent=TRUE)
imp <- as.data.frame(importance(richness_rfP))
imp1 <- imp[which(imp$MeanDecreaseAccuracy.pval < 0.05), c("MeanDecreaseAccuracy", "MeanDecreaseAccuracy.pval")]
confu <- confusionMatrix(richness_rfP)

## Single- vs co-
physeq.rpa<- subset_samples(physeq.rp, group2 %in% c("co-infection", "infection"))
physeq.rpa<-  prune_taxa(taxa_sums(physeq.rpa) > 0, physeq.rpa)
physeq10 = tax_glom(physeq.rpa, "genus")
data <- as.data.frame(t(physeq10@otu_table@.Data))
colnames(data) <- as.data.frame(physeq10@tax_table@.Data)$genus
colnames(data) <- make.names(colnames(data))
data.g <- data[,!colnames(data) %in% c("g__uncultured", "g__uncultured bacterium", "g__uncultured rumen bacterium","g__uncultured forest soil bacterium" ,"g__hydrothermal vent metagenome" ,"g__uncultured soil bacterium" ,"g__uncultured compost bacterium" ,"g__unidentified")]

colnames(data.g) <- make.names(colnames(data.g))

meta <- as.data.frame(meta(physeq10))
group1 <- subset(meta, group2 == "A")
group2 <- subset(meta, group2 != "A")
data1 <- data.g[rownames(group1),]
data2 <- data.g[sample(rownames(group2), nrow(group1)),]
data1$group <- "co-infection"
data2$group <- "Infection"
data.all <- rbind(data1, data2)
data.all$group <- as.factor(data.all$group)

set.seed(123)
richness_rfP<- rfPermute(group ~ ., data = data.all, ntree = 500,
                         na.action = na.omit, nrep = 100)
roc <- roc(data.all$group, richness_rfP$rf$votes[,1], percent=TRUE)
imp <- as.data.frame(importance(richness_rfP))
imp1 <- imp[which(imp$MeanDecreaseAccuracy.pval < 0.05), c("MeanDecreaseAccuracy", "MeanDecreaseAccuracy.pval")]
confu <- confusionMatrix(richness_rfP)

## A B S H in single
#   A   AB  ABH  ABS ABSH   AH   AS    B   BH   BS  BSH  CTL    H    S   SH 
#  22   16    4   12    2    3   13  138   22   16    3   98   61   34    3 

physeq.rpa<- subset_samples(physeq.rp, group != "NA")
physeq.rpa<-  prune_taxa(taxa_sums(physeq.rpa) > 0, physeq.rpa)
#physeq10 = tax_glom(physeq.rpa, "genus")
data <- as.data.frame(t(physeq10@otu_table@.Data))
#colnames(data) <- as.data.frame(physeq10@tax_table@.Data)$genus
#colnames(data) <- make.names(colnames(data))
#data.g <- data[,!colnames(data) %in% c("g__uncultured", "g__uncultured bacterium", "g__uncultured rumen bacterium","g__uncultured forest soil bacterium" ,"g__hydrothermal vent metagenome" ,"g__uncultured soil bacterium" ,"g__uncultured compost bacterium" ,"g__unidentified")]

colnames(data) <- make.names(colnames(data))
meta <- as.data.frame(meta(physeq.rpa))

#FluA
library(tidyverse)
library("randomForest")
library("rfUtilities")
library("rfPermute")

group1 <- subset(meta, group == "A")
group2 <- subset(meta, group != "A")
data1 <- data.g[rownames(group1),]
data2 <- data.g[sample(rownames(group2), nrow(group1)),]
data1$group <- "A"
data2$group <- "Others"
data.all <- rbind(data1, data2)
data.all$group <- as.factor(data.all$group)

set.seed(123)
richness_rfP<- rfPermute(group ~ ., data = data.all, ntree = 500,
                         na.action = na.omit, nrep = 100, num.cores = 6)
roc <- roc(data.all$group, richness_rfP$rf$votes[,1], percent=TRUE)
imp <- as.data.frame(importance(richness_rfP))
imp1 <- imp[which(imp$MeanDecreaseAccuracy.pval < 0.05), c("MeanDecreaseAccuracy", "MeanDecreaseAccuracy.pval")]
confu <- confusionMatrix(richness_rfP)

# FluB
group1 <- subset(meta, group == "B")
group2 <- subset(meta, group != "B")
data1 <- data.g[rownames(group2),]
data2 <- data.g[sample(rownames(group1), nrow(group2)),]
data2$group <- "FluB"
data1$group <- "Others"
data.all <- rbind(data1, data2)
data.all$group <- as.factor(data.all$group)

set.seed(123)
richness_rfP<- rfPermute(group ~ ., data = data.all, ntree = 500,
                         na.action = na.omit, nrep = 100)
roc <- roc(data.all$group, richness_rfP$rf$votes[,1], percent=TRUE)
imp <- as.data.frame(importance(richness_rfP))
imp1 <- imp[which(imp$MeanDecreaseAccuracy.pval < 0.05), c("MeanDecreaseAccuracy", "MeanDecreaseAccuracy.pval")]
confu <- confusionMatrix(richness_rfP)

# RSV
group1 <- subset(meta, group == "S")
group2 <- subset(meta, group != "S")
data1 <- data.g[rownames(group1),]
data2 <- data.g[sample(rownames(group2), nrow(group1)),]
data1$group <- "RSV"
data2$group <- "Others"
data.all <- rbind(data1, data2)
data.all$group <- as.factor(data.all$group)

set.seed(123)
richness_rfP<- rfPermute(group ~ ., data = data.all, ntree = 500,
                         na.action = na.omit, nrep = 100)
roc <- roc(data.all$group, richness_rfP$rf$votes[,1], percent=TRUE)
imp <- as.data.frame(importance(richness_rfP))
imp1 <- imp[which(imp$MeanDecreaseAccuracy.pval < 0.05), c("MeanDecreaseAccuracy", "MeanDecreaseAccuracy.pval")]
confu <- confusionMatrix(richness_rfP)

# RHV
group1 <- subset(meta, group == "H")
group2 <- subset(meta, group != "H")
data1 <- data.g[rownames(group1),]
data2 <- data.g[sample(rownames(group2), nrow(group1)),]
data1$group <- "RHV"
data2$group <- "Others"
data.all <- rbind(data1, data2)
data.all$group <- as.factor(data.all$group)

set.seed(123)
richness_rfP<- rfPermute(group ~ ., data = data.all, ntree = 500,
                         na.action = na.omit, nrep = 100)
roc <- roc(data.all$group, richness_rfP$rf$votes[,1], percent=TRUE)
imp <- as.data.frame(importance(richness_rfP))
imp1 <- imp[which(imp$MeanDecreaseAccuracy.pval < 0.05), c("MeanDecreaseAccuracy", "MeanDecreaseAccuracy.pval")]
confu <- confusionMatrix(richness_rfP)


library(tidyverse)
library(reshape2)

roc <- read.csv("C:/Users/DELL/Desktop/URTI.16SV4.202110/fig/FIG3/roc.csv",header = T, sep = ',', row.names = 1)
roc1 <- melt(roc)
data_m_sd_mean <- roc1 %>% group_by(variable) %>% dplyr::summarise(sd=sd(value), value=mean(value))
data_m_sd_mean <- as.data.frame(data_m_sd_mean)
data_m_sd_mean$variable <- factor(data_m_sd_mean$variable,levels=c("A", "B", "S", "H", "AB", "ABS", "AS", "BH", "BS", "infection", "co.infection", "CTL"))

col12<- c("#824533", "#B79570","#CAD09E", "#FFCCB1", "#161645",  "#2b2950", "#333d72", "#6b6bcf", "#9da3d7",  "#EC3D33", "#e3766b", "#2B6B8E")
ggplot(data_m_sd_mean, aes(x=variable, y=value)) + 
    geom_bar(stat="identity", aes(fill=variable)) +
    scale_fill_manual(values=col12) +
    geom_errorbar(aes(ymin=value-sd, ymax=value+sd), width=0.2, position=position_dodge(width=0.75)) +
    theme_classic()

imp <- read.csv("C:/Users/DELL/Desktop/URTI.16SV4.202110/fig/FIG3/imp.csv",header = T, sep = ',')
data <- as.data.frame(aggregate(imp$out_id, by=list(type=imp$Group_type,imp$out_id),length))
data.50 <- subset(data, x >= 50)
tax <- physeq.rpa@tax_table@.Data
tax1 <- tax[data.50$Group.2, ]
data2 <- subset(data, Group.2 %in% data.50$Group.2)
data3 <- dcast(data2, type ~ Group.2)
rownames(data3) <- data3$type
data3 <- data3[c("A", "B", "S", "H", "AB", "ABS", "AS", "BH", "BS", "infection", "co-infection", "CTL"),-1]
colnames(data3) <- paste(tax1[colnames(data3),]$Taxon, colnames(data3))
data3 <- data3[,colnames(data3)[order(colnames(data3))]]
pheatmap::pheatmap(t(data3),cluster_rows = FALSE,  cluster_cols = FALSE)

# plotImportance(richness_rfP)

# plotConfMat(richness_rfP)

 plot.roc(data.all$group, 
         richness_rfP$rf$votes[,1], 
         percent=TRUE,
         xlab="False Positive Percentage", 
         ylab="True Postive Percentage", col="#377eb8", lwd=4, print.auc=TRUE)
```


## sparcc
```{r}
library(Matrix)
library(car)
library(igraph)
library(RCy3)
library(tidyverse)

physeq.a<- subset_samples(physeq.rp, group != "NA")
physeq.ap<-  prune_taxa(taxa_sums(physeq.a) > 0, physeq.a)
physeq.ap.genus = tax_glom(physeq.ap, "genus")
otus <- as.data.frame(t(physeq.ap.genus@otu_table@.Data))
colnames(otus) <- as.data.frame(physeq.ap.genus@tax_table@.Data)$genus
meta <- as.data.frame(meta(physeq.a))
otus$group <- meta$group
otus1 <- otus[,!colnames(otus) %in% c("g__uncultured", "g__uncultured bacterium", "g__unidentified")]
colnames(otus1) <- make.names(colnames(otus1))

dat <- otus1 %>%
    group_by(group) %>%
    summarise_if(is.numeric, list(~mean(.)))
dat <- as.data.frame(dat)
rownames(dat) <- dat$group
dat<- dat[,-1]

tax <- read.csv("C:/Users/DELL/Desktop/URTI.16SV4.202110/sparcc.urti/tax.sparcc.csv",header = T, sep = ',',row.names = 1)

r.CTL<- read.table('C:/Users/DELL/Desktop/URTI.16SV4.202110/sparcc.urti/ResultsSparCC.CTL/CTL_sparcc.txt', header = T, sep = '\t', row.names = 1)
p.CTL<- read.table('C:/Users/DELL/Desktop/URTI.16SV4.202110/sparcc.urti/ResultsSparCC.CTL/CTL_pvals_two_sided.txt', header = T, sep = '\t', row.names = 1)

r.CTL[p.CTL>0.01] = 0
r.CTL1 <- as.matrix(r.CTL)

tax.CTL1 <- tax[rownames(r.CTL),]

igraph.CTL<-  graph_from_adjacency_matrix(r.CTL1,
                                        mode="undirected",
                                        weighted=TRUE,
                                        diag=FALSE )

V(igraph.CTL)$phylum <- tax.CTL1$phylum
V(igraph.CTL)$genus <- tax.CTL1$genus
tax.CTL1$mean <-  as.numeric(log10(dat["CTL",which(dat["CTL",] > 0)]/2+1))
V(igraph.CTL)$mean <- tax.CTL1$mean

createNetworkFromIgraph(igraph.CTL,"CTL")


```

## 鲁棒性1

```{r}
setwd("C:/Users/DELL/Desktop/URTI.16SV4.202110/sparcc.urti")

physeq.a<- subset_samples(physeq.rp, group == "A")
physeq.ap<-  prune_taxa(taxa_sums(physeq.a) > 0, physeq.a)
physeq.ap.genus = tax_glom(physeq.ap, "genus")
otus <- as.data.frame(t(physeq.ap.genus@otu_table@.Data))
colnames(otus) <- as.data.frame(physeq.ap.genus@tax_table@.Data)$genus
otus1 <- otus[,!colnames(otus) %in% c("g__uncultured", "g__uncultured bacterium", "g__unidentified")]
colnames(otus1) <- make.names(colnames(otus1))

sp.ra<-colMeans(otus1)/20000   #relative abundance of each species

###### there are two choices to get the correlation matrix #######


###### choice 2: directely read in correlation matrix downloaded from MENAP. MENAP downloaded correlation matrix is an upper triangle. Need to make it into a symetric matrix.
r.CTL<- read.table('C:/Users/DELL/Desktop/URTI.16SV4.202110/sparcc.urti/ResultsSparCC.A/A_sparcc.txt', header = T, sep = '\t', row.names = 1)
p.CTL<- read.table('C:/Users/DELL/Desktop/URTI.16SV4.202110/sparcc.urti/ResultsSparCC.A/A_pvals_two_sided.txt', header = T, sep = '\t', row.names = 1)
cormatrix <- r.CTL
cormatrix[p.CTL>0.01] = 0

###### end of the two choices of correlation matrix ########
names(sp.ra) <- colnames(cormatrix)
cormatrix2<-cormatrix*(abs(cormatrix)>=0)  #only keep links above the cutoff point
cormatrix2[is.na(cormatrix2)]<-0
diag(cormatrix2)<-0    #no links for self-self    
sum(abs(cormatrix2)>0)/2  #this should be the number of links. 
sum(colSums(abs(cormatrix2))>0)  # node number: number of species with at least one linkage with others.

network.raw<-cormatrix2[colSums(abs(cormatrix2))>0,colSums(abs(cormatrix2))>0]
sp.ra2<-sp.ra[colSums(abs(cormatrix2))>0]
sum(row.names(network.raw)==names(sp.ra2))  #check if matched

## robustness simulation 
#input network matrix, percentage of randomly removed species, and ra of all species
#return the proportion of species remained

rand.remov.once<-function(netRaw, rm.percent, sp.ra, abundance.weighted=T){
  id.rm<-sample(1:nrow(netRaw), round(nrow(netRaw)*rm.percent))
  net.Raw=netRaw #don't want change netRaw
  net.Raw[id.rm,]=0;  net.Raw[,id.rm]=0;   ##remove all the links to these species
  if (abundance.weighted){
    net.stength= net.Raw*sp.ra
  } else {
    net.stength= net.Raw
  }
 
  sp.meanInteration<-colMeans(net.stength)

  id.rm2<- which(sp.meanInteration<=0)  ##remove species have negative interaction or no interaction with others
  remain.percent<-(nrow(netRaw)-length(id.rm2))/nrow(netRaw)
  #for simplicity, I only consider the immediate effects of removing the
  #'id.rm' species; not consider the sequential effects of extinction of
  # the 'id.rm2' species.
  
  #you can write out the network pruned
  #  net.Raw[id.rm2,]=0;  net.Raw[,id.rm2]=0;
  # write.csv( net.Raw,"network pruned.csv")
  
  remain.percent
}

rm.p.list=seq(0.05,0.2,by=0.05)
rmsimu<-function(netRaw, rm.p.list, sp.ra, abundance.weighted=T,nperm=100){
  t(sapply(rm.p.list,function(x){
    remains=sapply(1:nperm,function(i){
      rand.remov.once(netRaw=netRaw, rm.percent=x, sp.ra=sp.ra, abundance.weighted=abundance.weighted)
    })
    remain.mean=mean(remains)
    remain.sd=sd(remains)
    remain.se=sd(remains)/(nperm^0.5)
    result<-c(remain.mean,remain.sd,remain.se)
    names(result)<-c("remain.mean","remain.sd","remain.se")
    result
  }))
}

Weighted.simu<-rmsimu(netRaw=network.raw, rm.p.list=seq(0.05,1,by=0.05), sp.ra=sp.ra2, abundance.weighted=T,nperm=100)
Unweighted.simu<-rmsimu(netRaw=network.raw, rm.p.list=seq(0.05,1,by=0.05), sp.ra=sp.ra2, abundance.weighted=F,nperm=100)

dat1<-data.frame(Proportion.removed=rep(seq(0.05,1,by=0.05),2),rbind(Weighted.simu,Unweighted.simu),
                 weighted=rep(c("weighted","unweighted"),each=20), treat=rep("A",40))


#write.csv(currentdat,"random_removal_result_Y14_W.csv")


##plot
library(ggplot2)

currentdat <- rbind(A, dat1, dat2,dat3, dat.s, dat.h, dat.abs, dat.as, dat.bh, dat.bs)
currentdat$treat <- factor(currentdat$treat,levels=c("A", "B", "S", "H", "AB", "ABS", "AS", "BH", "BS", "CTL"))
ggplot(currentdat[currentdat$weighted=="unweighted",], aes(x=Proportion.removed, y=remain.mean, group=treat, color=treat)) + 
    geom_line(size=1.5, alpha=0.8)+
    geom_pointrange(aes(ymin=remain.mean-remain.sd, ymax=remain.mean+remain.sd),size=0.5)+
    scale_color_manual(values=col10)+
    xlab("Proportion of species removed")+
    ylab("Proportion of species remained")+
    theme_classic()
  
```

## 鲁棒性2

```{r}
setwd("C:/Users/DELL/Desktop/URTI.16SV4.202110/sparcc.urti")

physeq.a<- subset_samples(physeq.rp, group == "A")
physeq.ap<-  prune_taxa(taxa_sums(physeq.a) > 0, physeq.a)
physeq.ap.genus = tax_glom(physeq.ap, "genus")
otus <- as.data.frame(t(physeq.ap.genus@otu_table@.Data))
colnames(otus) <- as.data.frame(physeq.ap.genus@tax_table@.Data)$genus
otus1 <- otus[,!colnames(otus) %in% c("g__uncultured", "g__uncultured bacterium", "g__unidentified")]
colnames(otus1) <- make.names(colnames(otus1))

sp.ra<-colMeans(otus1)/20000   #relative abundance of each species

###### there are two choices to get the correlation matrix #######
###### choice 1 (slow): calculate correlation matrix from OTU table

r.CTL<- read.table('C:/Users/DELL/Desktop/URTI.16SV4.202110/sparcc.urti/ResultsSparCC.A/A_sparcc.txt', header = T, sep = '\t', row.names = 1)
p.CTL<- read.table('C:/Users/DELL/Desktop/URTI.16SV4.202110/sparcc.urti/ResultsSparCC.A/A_pvals_two_sided.txt', header = T, sep = '\t', row.names = 1)
cormatrix <- r.CTL
cormatrix[p.CTL>0.01] = 0

###### end of the two choices of correlation matrix ########
names(sp.ra) <- colnames(cormatrix)
cormatrix2<-cormatrix*(abs(cormatrix)>=0)  #only keep links above the cutoff point
cormatrix2[is.na(cormatrix2)]<-0
diag(cormatrix2)<-0    #no links for self-self    
sum(abs(cormatrix2)>0)/2  #this should be the number of links. 
sum(colSums(abs(cormatrix2))>0)  # node number: number of species with at least one linkage with others.

network.raw<-cormatrix2[colSums(abs(cormatrix2))>0,colSums(abs(cormatrix2))>0]
sp.ra2<-sp.ra[colSums(abs(cormatrix2))>0]
sum(row.names(network.raw)==names(sp.ra2))  #check if matched
## robustness simulation 
#input network matrix, number of removed keystone species, keystonespecies list,  and ra of all species
#return the proportion of species remained

#get the keystone species list
r.CTL1 <- as.matrix(cormatrix)
igraph.CTL<-  graph_from_adjacency_matrix(r.CTL1,
                                        mode="undirected",
                                        weighted=TRUE,
                                        diag=FALSE )
library(ggClusterNet)
res <- ZiPiPlot(igraph = igraph.CTL, method = "cluster_fast_greedy")
node.attri <- as.data.frame(res[[2]])
module.hub<-rownames(node.attri)[node.attri$z > 2.5 & node.attri$p <= 0.62]

#consider cascade effects: removed species will further influence the remaining nodes

rand.remov2.once<-function(netRaw, rm.num, keystonelist, sp.ra, abundance.weighted=T){
  rm.num2<-ifelse(rm.num > length(keystonelist), length(keystonelist), rm.num)
  id.rm<-sample(keystonelist, rm.num2)
  net.Raw=netRaw #don't want change netRaw
  
  net.new=net.Raw[!names(sp.ra) %in% id.rm, !names(sp.ra) %in% id.rm]   ##remove all the links to these species
  if (nrow(net.new)<2){
    0
  } else {
    sp.ra.new=sp.ra[!names(sp.ra) %in% id.rm]
    
    if (abundance.weighted){
      net.stength= net.new*sp.ra.new
    } else {
      net.stength= net.new
    }
    
    sp.meanInteration<-colMeans(net.stength)
    
    
    while ( length(sp.meanInteration)>1 & min(sp.meanInteration) <=0){
      id.remain<- which(sp.meanInteration>0) 
      net.new=net.new[id.remain,id.remain]
      sp.ra.new=sp.ra.new[id.remain]
      
      if (abundance.weighted){
        net.stength= net.new*sp.ra.new
      } else {
        net.stength= net.new
      }
      
      if (length(net.stength)>1){
        sp.meanInteration<-colMeans(net.stength)
      } else{
        sp.meanInteration<-0
      }
      
    }
    
    remain.percent<-length(sp.ra.new)/length(sp.ra)
    
    remain.percent}
}

rm.p.list=seq(0.05,0.2,by=0.05)
rmsimu<-function(netRaw, rm.p.list, keystonelist,sp.ra, abundance.weighted=T,nperm=100){
  t(sapply(rm.p.list,function(x){
    remains=sapply(1:nperm,function(i){
      rand.remov2.once(netRaw=netRaw, rm.num=x, keystonelist=keystonelist, sp.ra=sp.ra, abundance.weighted=abundance.weighted)
    })
    remain.mean=mean(remains)
    remain.sd=sd(remains)
    remain.se=sd(remains)/(nperm^0.5)
    result<-c(remain.mean,remain.sd,remain.se)
    names(result)<-c("remain.mean","remain.sd","remain.se")
    result
  }))
}

Weighted.simu<-rmsimu(netRaw=network.raw, rm.p.list=1:length(module.hub),keystonelist=module.hub, sp.ra=sp.ra2, abundance.weighted=T,nperm=100)
Unweighted.simu<-rmsimu(netRaw=network.raw, rm.p.list=1:length(module.hub), keystonelist=module.hub, sp.ra=sp.ra2, abundance.weighted=F,nperm=100)

dat1<-data.frame(Number.hub.removed=rep(1:length(module.hub),2),rbind(Weighted.simu,Unweighted.simu),
                 weighted=rep(c("weighted","unweighted"),each=length(module.hub)),treat=rep("A",2*length(module.hub)))

currentdat = dat1
#write.csv(currentdat,"targeted_deletion_results_Y14_W.csv")

##plot
library(ggplot2)

currentdat$Number.hub.removed

ggplot(currentdat[currentdat$weighted=="unweighted",], aes(x=Number.hub.removed, y=remain.mean, group=treat, color=treat)) + 
    geom_line(size=1.5, alpha=0.8)+
    geom_pointrange(aes(ymin=remain.mean-remain.sd, ymax=remain.mean+remain.sd),size=0.5)+
    scale_color_manual(values=col10)+
    xlab("Number.hub.removed")+
    ylab("Proportion of species remained")+
    theme_classic()

```

## fig5

```{r}

grid <- read.csv("C:/Users/DELL/Desktop/URTI.16SV4.202110/fig/FIG5/grid.csv",header = T, sep = ',')


# names = c("Actinomyces graevenitzii","Capnocytophaga ochracea", "Fusobacterium nucleatum", "Haemophilus sputorum", "Neisseria flavescens", "Parvimonas micra", "Prevotella intermedia", "Prevotella sp oral taxon 306", "Prevotella oris", "Rothia aeria", "Rothia mucilaginosa", "Streptococcus infantis", "Streptococcus cristatus", "Streptococcus mitis oralis pneumoniae", "Streptococcus sp SK140", "Treponema medium")

col5<- c("#824533", "#FFCCB1", "#B79570", "#2B6B8E")

names1 <- c("Streptococcus pneumoniae", "Mycoplasma pneumoniae", "Haemophilus influenzae", "Klebsiella pneumoniae", "Pseudomonas aeruginosa", "Staphylococcus aureus", "Chlamydia pneumoniae", "Legionella pneumophila", "Escherichia coli", "Treponema denticola", "Porphyromonas gingivalis", "Haemophilus sputorum","Staphylococcus epidermidis", "Fusobacterium nucleatum", "Treponema medium")

data1 <- subset(grid, species %in% names1)


data1$group <- factor(data1$group,levels=c("FluA", "FluB", "FluAB", "Health"))
my_comparisons2 <- list(c("Health", "FluA"), c("Health", "FluB"), c("Health", "FluAB"))

ggplot(data=data1, aes(x = group, y = GRiD_unrefined, group=group)) +
    geom_boxplot(aes(x=group, fill=group), notch=FALSE, outlier.colour = "grey") +
    stat_summary(aes(x=group), fun.y=mean, geom="point", color="grey") +
    geom_jitter(aes(color=species), width=0.2, size=1.5, alpha = 0.75) +
    scale_fill_manual(values=col5) +
    theme(title=element_text(size=14,color="#4F4F4F"))+
    stat_compare_means(comparisons = my_comparisons2, paired = FALSE, color="black")+
    theme_classic()+ 
    theme(axis.text.x = element_text(size=14),axis.text.y = element_text(size=14)) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
    theme(legend.title = element_blank())+
    theme(plot.title = element_text(hjust = 0.5))

sp <- data1$species[!duplicated(data1$species)]

species <- read.csv("C:/Users/DELL/Desktop/URTI.16SV4.202110/fig/FIG5/species.csv",header = T, sep = ',', row.names = 1)

sp1 <- species["Treponema denticola", data1$id[which(data1$species == "Treponema denticola" )]]

data1 <- subset(grid, species %in% sp)
grid1 <- data1[, c(3,9,5)]
widedata2 <- dcast(grid1, id~species )
rownames(widedata2) <- widedata2$id
widedata2 <- widedata2[,-1]
path <- read.csv("C:/Users/DELL/Desktop/URTI.16SV4.202110/fig/FIG5/virulence.csv",header = T, sep = ',', row.names = 1)

path <- as.data.frame(t(path))
widedata2 <- widedata2[rownames(path),]

library(psych)
library(corrplot)
cor <- corr.test(widedata2, path, method = "spearman")
r <- cor$r
p <- cor$p
r[p > 0.05 | abs(r) < 0.5] = 0
r1 <- r[, which(colSums(abs(r)) > 0)]

col3 <- colorRampPalette(c("blue", "white", "red"))
corrplot(r1,col=col3(20),method='color',tl.pos='lt', tl.col='black', sig.level=c(.001, .01, .05), pch.cex= .9, pch.col='black', xpd = T, tl.srt = 45)

data1 <- subset(grid, species %in% names1)
data2 <- grid[-c(which(grid$species %in% names1)),]
test1 <- dcast(data1,id~data1$species,value.var = 'GRiD_unrefined')
rownames(test1) <- test1$id
test1 <- test1[,-1]
test1 =  test1[, apply(test1, 2, function(y) length(na.omit(y))>5)]
test2 <- dcast(data2,id~data2$species,value.var = 'GRiD_unrefined')
rownames(test2) <- test2$id
test2 <- test2[,-1]
test2 =  test2[, apply(test2, 2, function(y) length(na.omit(y))>5)]

cor <- corr.test(test1, test2, method = "spearman")
r <- cor$r
p <- cor$p

pheatmap(r, cluster_rows=T, cluster_cols=T)

r[p > 0.05] = 0
r1 <- r[, which(colSums(abs(r)) > 0)]


col3 <- colorRampPalette(c("blue", "white", "red"))
corrplot(r1,col=col3(20),method='color',tl.pos='lt', tl.col='black', sig.level=c(.001, .01, .05), pch.cex= .9, pch.col='black', xpd = T, tl.srt = 45)

```